client.go

```diff
@@ -238,6 +251,11 @@ func (client *HTTPClient) perform(method, url string, entity interface{}, params
 			if err := client.do(req); err != nil {
 				return err
 			}
+			defer req.RawResponse.Body.Close()
+
+			if !client.retryPredicate(req, err) {
```

Comment:
Wenn das Prädikat den Body liest, dann schlägt in den meisten Fällen der Aufruf von `req.decodeBody(entity)` fehl. Da der `io.ReadCloser` nur einmal gelesen werden kann. Das könnte man nur umgehen, wenn man den Body buffered.

Das Predikat sollte besser auch nur ausgewertet werden, wenn der Client wirklich einen Retry in Erwägung zieht. Z.b. hier https://github.com/snabble/go-httputils/blob/e12c766c0b81b5d03560a6abff73c91ebcebeb38/client.go#L470

Reply:
Mhh, das Problem was ich habe ist, dass der aufrufende Code davon ausgeht das decode body relativ früh im processing aufgerufen wird, also entweder saugt mir decode den body leer oder mein predicate. Ich kann nur die Funktionsweise vom decodeBody so wie im Vorschlag Chat umbauen.

Auf Zeile 470 landet der GET Code aber auch nie.